Add support for dart in vscode-dart

We are working on improving graphql support for dart. vscode-apollo is the best plugin to provide syntax highlighting in vscode and we wouldn't want to create a separate plugin just to allow syntax highlighting in dart.

This would just be a matter of adding a new template syntax to vscode-apollo. We can use this as a reference implementation: https://github.com/apollographql/apollo-tooling/blob/master/packages/vscode-apollo/syntaxes/graphql.py.json

If the Apollo team is OK with this, I or someone on my team can submit a pull request with the additional syntax.

The syntax we would support will look like this:
```dart
var query = gql("""
                query GetCard(\$id:CardId!) {
                  card(id:\$id) {
                    id
                    name
                    completed
                }
                """);
```

Hi @JakeDawkins I'm getting the following error when I follow those steps and run the extension:

```Debugger listening on ws://127.0.0.1:6009/13cecaea-ae2b-40eb-9dcb-805162c06d36
For help, see: https://nodejs.org/en/docs/inspector
internal/modules/cjs/loader.js:586
    throw err;
    ^

Error: Cannot find module 'apollo-env'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:584:15)
    at Function.Module._load (internal/modules/cjs/loader.js:510:25)
    at Module.require (internal/modules/cjs/loader.js:640:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at Object.<anonymous> (/Users/venkat/code/apollo-tooling/packages/apollo-language-server/lib/server.js:3:1)
    at Object.<anonymous> (/Users/venkat/code/apollo-tooling/packages/apollo-language-server/lib/server.js:119:3)
    at Module._compile (internal/modules/cjs/loader.js:693:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:704:10)
    at Module.load (internal/modules/cjs/loader.js:602:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:541:12)
[Info  - 3:42:19 PM] Connection to server got closed. Server will restart.
Debugger listening on ws://127.0.0.1:6009/b4d2c356-c5c2-4914-bf74-6880423c20bf
For help, see: https://nodejs.org/en/docs/inspector
internal/modules/cjs/loader.js:586
    throw err;
    ^
```

Any idea why this might be the case? The `package.json` refers to `apollo-env` by relative path not sure if this has anything to do with it.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,11 +19,11 @@
 - `apollo-graphql`
   - <First `apollo-graphql` related entry goes here>
 - `apollo-language-server`
-  - <First `apollo-language-server` related entry goes here>
+  - Add Dart operation extraction [#1385](https://github.com/apollographql/apollo-tooling/pull/1385)
 - `apollo-tools`
   - <First `apollo-tools` related entry goes here>
 - `vscode-apollo`
-  - <First `vscode-apollo` related entry goes here>
+  - Add Dart support for vscode [#1385](https://github.com/apollographql/apollo-tooling/pull/1385)
 
 ## `apollo@2.16.1`, `apollo-language-server@1.13.1`, `vscode-apollo@1.8.1`
 

--- a/packages/apollo-language-server/src/document.ts
+++ b/packages/apollo-language-server/src/document.ts
@@ -70,6 +70,8 @@ export function extractGraphQLDocuments(
       return extractGraphQLDocumentsFromPythonStrings(document, tagName);
     case "ruby":
       return extractGraphQLDocumentsFromRubyStrings(document, tagName);
+    case "dart":
+      return extractGraphQLDocumentsFromDartStrings(document, tagName);
     default:
       return null;
   }
@@ -145,6 +147,36 @@ function extractGraphQLDocumentsFromRubyStrings(
   let result;
   while ((result = regExp.exec(text)) !== null) {
     const contents = replacePlaceholdersWithWhiteSpace(result[2]);
+    const position = document.positionAt(result.index + result[1].length);
+    const locationOffset: SourceLocation = {
+      line: position.line + 1,
+      column: position.character + 1
+    };
+    const source = new Source(contents, document.uri, locationOffset);
+    documents.push(new GraphQLDocument(source));
+  }
+
+  if (documents.length < 1) return null;
+
+  return documents;
+}
+
+function extractGraphQLDocumentsFromDartStrings(
+  document: TextDocument,
+  tagName: string
+): GraphQLDocument[] | null {
+  const text = document.getText();
+
+  const documents: GraphQLDocument[] = [];
+
+  const regExp = new RegExp(
+    `\\b(${tagName}\\(\\s*r?("""|'''))([\\s\\S]+?)\\2\\s*\\)`,
+    "gm"
+  );
+
+  let result;
+  while ((result = regExp.exec(text)) !== null) {
+    const contents = replacePlaceholdersWithWhiteSpace(result[3]);
     const position = document.positionAt(result.index + result[1].length);
     const locationOffset: SourceLocation = {
       line: position.line + 1,

--- a/packages/apollo-language-server/src/project/base.ts
+++ b/packages/apollo-language-server/src/project/base.ts
@@ -41,7 +41,8 @@ const fileAssociations: { [extension: string]: string } = {
   ".tsx": "typescriptreact",
   ".vue": "vue",
   ".py": "python",
-  ".rb": "ruby"
+  ".rb": "ruby",
+  ".dart": "dart"
 };
 
 export interface GraphQLProjectConfig {

--- a/packages/vscode-apollo/package.json
+++ b/packages/vscode-apollo/package.json
@@ -110,6 +110,16 @@
         "embeddedLanguages": {
           "meta.embedded.block.graphql": "graphql"
         }
+      },
+      {
+        "injectTo": [
+          "source.dart"
+        ],
+        "scopeName": "inline.graphql.dart",
+        "path": "./syntaxes/graphql.dart.json",
+        "embeddedLanguages": {
+          "meta.embedded.block.graphql": "graphql"
+        }
       }
     ],
     "commands": [

--- a/packages/vscode-apollo/src/languageServerClient.ts
+++ b/packages/vscode-apollo/src/languageServerClient.ts
@@ -48,13 +48,14 @@ export function getLanguageServerClient(
       "typescriptreact",
       "vue",
       "python",
-      "ruby"
+      "ruby",
+      "dart"
     ],
     synchronize: {
       fileEvents: [
         workspace.createFileSystemWatcher("**/.env"),
         workspace.createFileSystemWatcher(
-          "**/*.{graphql,js,ts,jsx,tsx,vue,py,rb}"
+          "**/*.{graphql,js,ts,jsx,tsx,vue,py,rb,dart}"
         )
       ]
     },

--- a/packages/vscode-apollo/syntaxes/graphql.dart.json
+++ b/packages/vscode-apollo/syntaxes/graphql.dart.json
@@ -1,0 +1,40 @@
+{
+  "fileTypes": [
+    "dart"
+  ],
+  "injectionSelector": "L:source -string -comment",
+  "patterns": [
+    {
+      "name": "meta.function-call.dart",
+      "begin": "\\b(gql)(\\()",
+      "beginCaptures": {
+        "1": {
+          "name": "entity.name.function.dart"
+        },
+        "2": {
+          "name": "punctuation.definition.arguments.begin.dart"
+        }
+      },
+      "end": "(\\))",
+      "endCaptures": {
+        "1": {
+          "name": "punctuation.definition.arguments.end.dart"
+        }
+      },
+      "patterns": [
+        {
+          "name": "taggedTemplates",
+          "contentName": "meta.embedded.block.graphql",
+          "begin": "r?(\"\"\"|''')",
+          "end": "((\\1))",
+          "patterns": [
+            {
+              "include": "source.graphql"
+            }
+          ]
+        }
+      ]
+    }
+  ],
+  "scopeName": "inline.graphql.dart"
+}

